README.md
# qtMUD

[![Join the chat at https://gitter.im/qtmud/Lobby](https://badges.gitter.im/qtmud/Lobby.svg)](https://gitter.im/qtmud/Lobby?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![PyPI version](https://badge.fury.io/py/qtmud.svg)](https://badge.fury.io/py/qtmud)
[![Documentation Status](https://readthedocs.org/projects/qtmud/badge/?version=latest)](http://qtmud.readthedocs.io/en/latest/?badge=latest)

[![Code Climate](https://codeclimate.com/github/emsenn/qtmud/badges/gpa.svg)](https://codeclimate.com/github/emsenn/qtmud)
[![Build Status](https://travis-ci.org/emsenn/qtmud.svg?branch=master)](https://travis-ci.org/emsenn/qtmud)

[![Requirements Status](https://requires.io/github/emsenn/qtmud/requirements.svg?branch=master)](https://requires.io/github/emsenn/qtmud/requirements/?branch=master)

------

qtMUD is a Python3 package for developing and hosting MUDs, Multi-User 
Dimensions.

Complete documentation is available at 
[ReadTheDocs](http://qtmud.readthedocs.io/en/latest/).


## Installation

```bash
$ pip install qtmud
```

## Usage

The qtMUD package comes with `qtmud_run`, which runs a socket server.


```bash
$ qtmud_run -h
usage: qtmud_run [-h] [-v | -q] [--conf CONF]

optional arguments:
  -h, --help     show this help message and exit
  -v, --verbose  show all logging messages
  -q, --quiet    show only warning & more severe messages
  --conf CONF    config file to use

$ qtmud_run
qtmud_run preparing to start qtmud 0.0.5
qtmud        INFO     qtmud.load() called
qtmud        INFO     adding qtmud.subscriptions to qtmud.subscribers
qtmud        INFO     adding qtmud.services to qtmud.active_services
qtmud        INFO     qtmud.load()ed
qtmud        INFO     filling qtmud.client_accounts from ./qtmud_client_accounts.p
qtmud        INFO     start()ing active_services
qtmud        INFO     talker start()ed
qtmud        INFO     start()ing MUDSocket
qtmud        INFO     MUDSocket successfully bound to ('0.0.0.0', 5787)
qtmud        INFO     MUDSocket successfully bound to ('localhost', 5788, 0, 0)
qtmud        INFO     mudsocket start()ed
qtmud        INFO     qtmud.run()ning
```
`KeyboardInterrupt` to end the `run()`ning process.

---

To connect as a client to the running socket server:

```bash
$ telnet localhost 5787
Trying ::1...
Connection failed: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.

qtmud               0.0.5

Successfully connected to qtmud, press enter to continue login...
```
While telnet is the easiest way to connect, we recommend the 
[tiny-fugue](https://github.com/kruton/tinyfugue) MUD client.

docs/changelog.rst
Changelog
=========

%%version%% (unreleased)
------------------------

- Merge branch 'release/0.0.8' into develop. [emsenn]

0.0.8 (2016-10-02)
------------------

Changes
~~~~~~~

- Added Development Flow to Development Docs. [emsenn]

- Documented Parameters in ``qtmud.cmds`` [emsenn]

- Added Tests for qtmud.cmds. [emsenn]

  Also preparing to use gitchangelog to aid with versioning moving
  forward.

Other
~~~~~

- Merge branch 'master' into develop. [emsenn]

- Cleaner Shutdown/Startup. [emsenn]

  Added --logdir and --datadir options to `./bin/qtmud_run`

  Also fixed bugs #21 and #23

- Codecov YML Addition. [emsenn]

- Merge branch 'master' of github.com:emsenn/qtmud. [emsenn]

- Update .travis.yml. [emsenn]

- Update .travis.yml. [emsenn]

- Start of qtmud.cmds Test Cases. [emsenn]

- Added Code Climate Integration to Travis CI. [emsenn]

- Merge branch 'master' of github.com:emsenn/qtmud. [emsenn]

- Delete environment.pickle. [emsenn]

- Merge branch 'release/0.0.7' [emsenn]

- Merge branch 'release/0.0.6' [emsenn]

  First real package, I think? I hope. I'm sorry to anyone reading this
  from the future, I am a newbie.

- Merge branch 'release/0.0.7' into develop. [emsenn]

0.0.7 (2016-10-01)
------------------

- V0.0.7: Doc Refactor & client_disconnect Fix. [emsenn]

- Doc Refactoring, Fixed client_disconnect Bug. [emsenn]

- Merge branch 'release/0.0.6' into develop. [emsenn]

0.0.6 (2016-10-01)
------------------

- First "Good" Packaging Release. [emsenn]

- Merge branch 'master' into release/0.0.6. [emsenn]

  I messed up the git flow whoops

- Some random cleanups. [emsenn]

- Documentation Update - Removed mudlib docs. [emsenn]

- Fixed Broken MUDSocket Test. [emsenn]

- Update README.md. [emsenn]

- Update README.md. [emsenn]

- Update README.md. [emsenn]

- Update README.md. [emsenn]

- Update README.md. [emsenn]

- Update README.md. [emsenn]

- Added CodeClimate Badge to README.md. [emsenn]

- Transition to Package. [emsenn]

  Shifting qtMUD into being an acceptable Python package.

- Code Coverage? [emsenn]

- Fixed ./data/ not being in git. [emsenn]

- More Cards. [emsenn]

  And a few other trivial changes.

- More Fireside Cards. [emsenn]

  Cleaning up of Fireside code, too

- Require sudo in Travis CI. [emsenn]

  Travis-CI doesn't like that we require brlapi, and while I could just
  remove the requirement, we are going to need it eventually.

- Added requirements.txt. [emsenn]

- Merge branch 'master' of github.com:emsenn/qtmud. [emsenn]

- Update README.md. [emsenn]

- Travis CI Scripts. [emsenn]

  Simple tests for Travis CI? Maybe? I don't get how it works.

- Fireside Documentation. [emsenn]

  Built and added Fireside documentation.

- Attempted Fix for ReadTheDocs error. [emsenn]

  error was Could not import extension sphinx.ext.githubpages (exception:
  No module named githubpages)

  this is what google said would help

- PEP8 Updates & Fireside Cards. [emsenn]

- Fireside Mudlib. [emsenn]

  Simple cardgame mudlib and some edits to qtmud methods

- Basic Talker Service. [emsenn]

  A really basic and lazy implementation of a talker service.

- Reduction of Dependence on Starhopper. [emsenn]

  qtMUD, through refactoring, became dependent on Starhopper methods.

  This fixes some of that./

- Documentation Hotfix Part Three. [emsenn]

- Merge branch 'master' of github.com:emsenn/qtmud. [emsenn]

- Merge pull request #16 from emsenn/develop. [emsenn]

  Develop

- Merge pull request #14 from emsenn/develop. [emsenn]

  Develop

- Merge branch 'release/0.0.4' [emsenn]

- Merge branch 'release/0.0.3' [emsenn]

  Release of version 0.0.3 to master woooo

- Merge branch 'release/0.0.4' into develop. [emsenn]

0.0.4 (2016-09-26)
------------------

- Bump to version 0.0.4. [emsenn]

- .gitignore hotfix. [emsenn]

- Documentation Hotfix. [emsenn]

- Documentation Cleanup, Separating Client and Ship in Starhopper.
  [emsenn]

  Title about says it all.

- Mirrored Starhopper Structure in Qtmud. [emsenn]

  Updated qtmud to use a package structure more in line with the updated
  starhopper structure.

- Bringing Back Documentation. [emsenn]

  It's back! and less messy than ever!

- Deleted Broken Documentation, Refactored Starhopper. [emsenn]

  The documentation wasn't rendering right so I just got rid of it.

  also, refactored starhopper. Need to shuffle qtmud to match,
  unfortunately. New system is way better, though.

- Merge branch 'feature/diceroller' into develop. [emsenn]

  Got a little carried away with this feature

- Starhopper Update. [emsenn]

  Got frustrated with trying to buy a full MMORPG in one go so made a
  dinky little space adventure game.

- Migration to Game Library. [emsenn]

  I realized a lot of stuff was in qtmud that was better suited for the
  specific libraries - not every game that gets built is going to want a
  "say" command, for instance.

- Refactor. [emsenn]

  It finally clicked with me what people were saying about organizing the
  engine differently, so this is me shuffling around toward doing that.

  A lot of functionality is broken but I like the new direction.

- Changed Thing's search methods, restructured lib. [emsenn]

  I know it looks like a lot of changes but it's really not much.

- Swordsmanship, Healthful, Acting Qualities in Lib. [emsenn]

  A few qualities to make use of the diceroller.

  Not pleased with any of this code but it's better than nothing.

- Merge branch 'feature/noise' into develop. [emsenn]

  Noises basically work, even if their trigger mechanism is a bit simple.

- Failed to add changes to last commit. [emsenn]

  Whooops!

- Fixed Issue #9 & Added Documentation Theme. [emsenn]

  Fixed Issue #9, where clients weren't removed from their location
  when they disconnect.

  Also, added cute little Tumblebeasts to the documentation!

- Additions to Library: Ye Olde Tavern. [emsenn]

  made ye olde tavern less of a filler thing and more of a real thing.


qtmud/__init__.py
""" The main qtmud module houses some constants, the Thing class, and methods
    for start()ing and tick()ing the MUD engine.
"""

#
import logging
import pickle
import types
import uuid
from inspect import getmembers, isfunction, isclass
from logging.config import dictConfig

from clint.textui import colored

from qtmud import cmds, services, subscriptions, txt

# GLOBAL REFERENCES
NAME = 'qtMUD'
""" Name of the MUD engine. """
__version__ = '0.0.11'
""" MUD engine version """
__url__ = 'https://qtmud.readthedocs.io/en/latest/'




IPv4_HOSTNAME = '0.0.0.0'
IPv4_MUDPORT = 5787
IPv6_HOSTNAME = 'localhost'
IPv6_MUDPORT = 5788
DATA_DIR = './data/'
LOG_DIR = './logs/'



MUDLIB = None

SPLASH = txt.SPLASH.format(**locals())
""" The text new connections see. """

client_accounts = dict()
""" Populated by load_client_accounts(), a list of client accounts stored by
registered name."""

events = dict()
""" Events scheduled to occur next tick, populated by :func:`schedule`."""
things = dict()
""" all the things that new_thing() has made"""
subscribers = dict()
""" Methods which will be called for any relevant :attr:`events`. Populated
by :func:`qtmud.load` to contain bound methods of the methods in
:mod:`qtmud.subscriptions` referenced by method name. """
active_services = dict()
""" Services which will have their tick() function called by
:func:`qtmud.tick`. Populated by :func:`qtmud.load` to contain instances of
the classes in :mod:`qtmud.services` referenced by class name. """
connected_clients = list()


class TalkerHandler(logging.Handler):
    def __init__(self):
        super(TalkerHandler, self).__init__()
        self.talker = active_services['talker']

    def emit(self, record):
        self.talker.broadcast(channel=record.levelname, speaker=record.name,
                              message=record.msg)


logging_config = dict(version=1,
                      formatters={'f': {'format': ('%(asctime)s %(name)-12s '
                                                   '%(levelname)-8s '
                                                   '%(message)s')}}, handlers={
        'stream': {'class': 'logging.StreamHandler', 'formatter': 'f',
                   'level': logging.DEBUG},
        'file': {'class': 'logging.FileHandler', 'filename': './debug.log',
                 'formatter': 'f', 'level': logging.DEBUG}},
                      root={'handlers': ['stream', 'file'],
                            'level': logging.DEBUG})
logging.config.dictConfig(logging_config)
log = logging.getLogger(__name__)


def load():
    """ Most importantly, puts every function from
    :mod:`qtmud.subscriptions` and every class from :mod:`qtmud.services`
    into :attr:`subscribers` and :attr:`active_services`, respectively.

    After populating `subscribers` and `active_services`, calls the start()
    method for each class in `active_services`, then attempts to load
    :attr:`MUDLIB`, if one is specified, and finally calls
    :func:`load_client_accounts`.
    """
    global active_services
    global subscribers
    log.info('qtmud.load() called')
    log.info('adding qtmud.subscriptions to qtmud.subscribers')
    subscribers = {s[1].__name__: [s[1]] for
                   s in getmembers(subscriptions) if isfunction(s[1])}
    log.info('adding qtmud.services to qtmud.active_services')
    active_services = {t[1].__name__.lower(): t[1]() for
                       t in getmembers(services) if isclass(t[1])}
    if 'talker' in active_services:
        log.addHandler(TalkerHandler())
        for level in ['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL']:
            active_services['talker'].new_channel(level)
    if load_client_accounts():
        log.debug('qtmud.client_accounts populated from '
                  'qtmud.load_client_accounts()')
    log.info('qtmud.load()ed')
    log.debug('subscribers are: %s', ', '.join(subscribers))
    log.debug('active_services are: %s',
              ', '.join([s for s in active_services]))
    return True


def load_client_accounts(file=(DATA_DIR + 'qtmud_client_accounts.p')):
    """ Populates :attr:`qtmud.client_accounts` with the pickle file
    specified. """
    global client_accounts
    log.info('filling qtmud.client_accounts from %s', file)
    try:
        client_accounts = pickle.load(open(file, 'rb'))
        log.debug('qtmud.client_accounts filled from %s', file)
        return True
    except FileNotFoundError:
        log.debug('no save file found, making one at %s', file)
        save_client_accounts()
        return False


def new_client_account(name, password, birthtime=None):
    """ Create a new client account in :attr:`client_accounts`"""
    client_accounts[name.lower()] = {'name': name,
                                     'password': password}
    if birthtime:
        client_accounts[name]['birthtime'] = birthtime
    log.debug('made the new client %s', name)
    save_client_accounts()
    return client_accounts[name]

def embolden(text):
    for marker in ['**', '*', '``', '`']:
        text = text.split(marker)
        for word in text[1::2]:
            if word:
                text[text.index(word)] = '%^B_WHITE%^' + word + '%^'
        text = ''.join(text)
    return text


def pinkfish_parse(text, requester=None):
    colors = {'BLACK': colored.black,
              'RED': colored.red,
              'GREEN': colored.green,
              'YELLOW': colored.yellow,
              'BLUE': colored.blue,
              'MAGENTA': colored.magenta,
              'CYAN': colored.cyan,
              'WHITE': colored.white}
    debug = ''
    working_text = ''
    layers = []
    bold = False
    delimiter = '%^'
    split_text = [c for c in embolden(text).split(delimiter)]
    if len(split_text) <= 1:
        return text
    position = 0
    while position < len(split_text):
        chunk = split_text[position]
        if chunk in ['B_'+c for c in colors]:
            bold=True
            chunk = ''.join(chunk.split('_')[1:])
        if chunk in colors:
            layers.append((chunk, bold))
            position += 1
        elif layers:
            layer = layers[-1]
            working_text += colors[layer[0]](chunk, bold=layer[1])+'\033[0;0m'
            position += 1
            if position < len(split_text):
                if split_text[position] not in colors and \
                                split_text[position] not in ['B_' + c for c in
                                                             colors]:
                    layers.pop(-1)
        else:
            working_text += chunk
            position += 1
    return working_text+'\033[0;0m'


def run():
    """ main loop """
    log.info('qtmud.run()ning')
    try:
        while True:
            tick()
    except KeyboardInterrupt:
        log.info('shutdown started')
        schedule('shutdown')
        tick()


def save_client_accounts(file=(DATA_DIR + 'qtmud_client_accounts.p')):
    """ Saves a pickle of the current :attr:`client_accounts`."""
    log.debug('saving client_accounts to %s', file)
    try:
        pickle.dump(client_accounts, open(file, 'wb'))
        return True
    except FileNotFoundError as err:
        log.warning('failed to save client_accounts: %s', err)


def schedule(sub, **payload):
    """ Schedules a call to sub with payload passed as parameters.
    """
    if not subscribers.get(sub, []):
        log.warning('Tried to schedule a %s event, no subscribers '
                    'listen to it though.', sub)
        return False
    for method in subscribers.get(sub, []):
        if method not in events:
            events[method] = []
        events[method].append(payload)
    return True


def search_connected_clients_by_name(name):
    """ Searches through :attr:`connected_clients` for one with a matching
    name. *(Ignores case)*

        :param name:        The name of the client you're looking for.
        :return list:       A list of connected clients with that name.
                            (Should probably just have one element.)
    """
    return [connected_clients[connected_clients.index(c)] for c in
            connected_clients
            if hasattr(c, 'name') and c.name.lower() == name.lower()]


def search_client_accounts_by_name(name):
    """ Searches through the :attr:`client_accounts` for a client with a
    matching name. *(For convenience, ignores case)*

        :param name:        The name of the client you're looking for
        :return list:       A list of the clients with that name. (Should
                            probably just have one element.)
    """
    return [c for c in client_accounts.keys() if c.lower() == name.lower()]


def start():
    """ Calls the start() function in every service in
    :attr:`active_services`.
    """
    log.info('start()ing active_services')
    for service in active_services:
        log.info('%s start()ing', service)
        try:
            active_services[service].start()
            log.info('%s start()ed', service)
        except AttributeError:
            log.info('%s has no start()', service)
        except RuntimeWarning as warning:
            log.warning('%s failed to start: %s', service, warning)
    return True


def tick():
    """ Processes the events in :attr:`current_events`, sending them out to
    the :attr:`subscribers` who are listening for them.

        Also calls tick() in every service in :attr:`active_services`
    """
    global events
    if events:
        current_events = events
        events = dict()
        for event in current_events:
            for call in current_events[event]:
                try:
                    event(**call)
                except Exception as err:
                    log.warning(err, exc_info=True)
    for service in active_services:
        try:
            active_services[service].tick()
        except AttributeError:
            pass
    return True


class Thing(object):
    """ Most objects clients interact with are Things

        Created with :func:`new_thing`, things are objects with a few
        attributes added on, mostly for enabling in-game reference of the
        objects.
    """

    def __init__(self, **kwargs):
        self._name = str()
        while True:
            self.identity = uuid.uuid4()
            if self.identity not in things.keys():
                break
        things[self.identity] = self
        """ Passed by :func:`new_thing`, `identity` is stored as a UUID """
        self.nouns = {'thing'}
        """ `nouns` represent lower-case nouns which may be used to reference
            the thing.
        """
        self.name = str(self.identity)
        self.adjectives = set()
        self.qualities = []
        self.update(kwargs)
        return

    @property
    def name(self):
        """ Properly-cased full name of a thing
            Any name a thing is given is also added to :attr:`Thing.nouns`,
            with the old name being removed. Same for adjectives.

            .. warning:: This has some wonkiness, in that "Eric Baez" can be
                         referred to as "Eric Baez" or "Baez" but not "Eric".
                         "Eric Thing" would work, though.
        """
        return self._name

    @name.setter
    def name(self, value):
        old_name = self._name.lower().split(' ')
        if len(old_name) > 1:
            old_adjectives = old_name[0:-1]
            for adjective in old_adjectives:
                try:
                    self.adjectives.remove(adjective)
                except KeyError:
                    pass
        try:
            self.nouns.remove(old_name[-1])
        except KeyError:
            pass
        new_name = value.lower().split(' ')
        self.nouns.add(new_name[-1])
        if len(new_name) > 1:
            adjectives = new_name[0:-1]
            for adjective in adjectives:
                self.adjectives.add(adjective)
        self.nouns.add(new_name[-1])
        self._name = value

    def update(self, attributes):
        """ Update multiple attributes of Thing at once.

            Example:
                >>> foo = Thing()
                >>> foo.update({'name': 'eric'})
                >>> foo.name
                eric

            .. warning:: Does not properly recognize setters
        """
        # todo account for custom setters
        for attribute, value in attributes.items():
            if hasattr(self, attribute):
                self.__dict__[attribute] = value
        return True


class Client(Thing):
    """ The thing which represents a client within qtmud.
    """

    def __init__(self, **kwargs):
        super(Client, self).__init__(**kwargs)
        self.addr = tuple()
        """ The client's address, represented by IP and port.
            .. warning:: Probably broken if you try and connect through IPv6
        """
        self.commands = {}
        """ A dict where keys will be the names of methods in :mod:`qtmud.cmds`
        and the value will be a pointer to that method.
        """
        for command, function in [m for m in getmembers(cmds) if
                                  isfunction(m[1])]:
            self.commands[command] = types.MethodType(function, self)
        self.input_parser = 'client_command_parser'
        """ The subscription which will be called when this client inputs a
        command. If you wish to overwrite the client's default command
        parser, you'll need to overwrite this."""
        self.send_buffer = str()
        """ The string representing data the client has received. Usually
        emptied each :func:`tick <qtmud.tick>` by
        :class:`MUDSocket <qtmud.services.MUDSocket>`"""
        self.recv_buffer = str()
        self.channels = list()
        """ All the :class:`Talker <qtmud.services.Talker>` channels the
        client is listening to. """